internal/gcs/main.go

```diff
@@ -0,0 +1,76 @@
+// Package gcs provides a gRPC server implementation for the Blob Storage Interface (BSI).
+// Google Cloud Storage (GCS) is the implemented storage service.
+package gcs
+
+import (
+	"context"
+	"log"
+	"net"
+	"time"
+
+	bsi "github.com/substratusai/substratus/internal/bsi"
+
+	storage "cloud.google.com/go/storage"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// server implements the bsi.ControllerServer interface.
+type server struct {
+	bsi.UnimplementedControllerServer
+	storageClient *storage.Client
+}
+
+// CreateSignedURL generates a signed URL for a specified GCS bucket and object path.
+func (s *server) CreateSignedURL(ctx context.Context, req *bsi.CreateSignedURLRequest) (*bsi.CreateSignedURLResponse, error) {
+	// Use the bucket and object path from the request
+	bucket := s.storageClient.Bucket(req.GetBucketName())
+	obj := bucket.Object(req.GetObjectPath())
+	if _, err := obj.Attrs(ctx); err != nil {
+		if err != storage.ErrObjectNotExist {
+			// An error occurred that was NOT ErrObjectNotExist.
+			// This is an unexpected error and we should return it.
+			return nil, err
+		}
+		// the object doesn't exist and we want to continue and create the signed URL.
+	}
+
+	// Create a signed URL
+	url, err := storage.SignedURL(req.GetBucketName(), req.GetObjectPath(), &storage.SignedURLOptions{
+		Method:  "GET",
+		Expires: time.Now().Add(time.Duration(req.GetExpirationSeconds()) * time.Second),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Format the completion time to the proto spec
+	completionTime := timestamppb.New(time.Now())
```

Comment:
Why include completion time as a response value?

Reply:
I just took a gander as the csi interface and they dont have this... I vote to remove it unless your have a compelling reason for keeping it.

Reply:
I think the use case Brandon was thinking of is to know if the signed URL you get isn't expired already? I am not sure that's required though.

Reply:
I don't think I have a great use case. If too short a window is requested by a client, they should have a reasonable idea of when the expiration is because it issued the request. Clients can still implement logic to handle retry under an assumed expiration in those case.